File: webapp-rust/src/appconfig.rs
use std::env;
use std::error::Error;

#[derive(Debug, Clone)]
pub struct AppConfig {
    pub loglevel: String,
    pub host: String,
    pub port: u16,
    pub existdb_uri: String,
    pub auth_userheader: String,
    pub dev_authuser: Option<String>,
}

pub fn read_appconfig() -> Result<AppConfig, Box<dyn Error>> {
    Ok(AppConfig {
        loglevel: env::var("LOGLEVEL").unwrap_or_else(|_| String::from("info")),
        host: env::var("HOST").unwrap_or_else(|_| String::from("0.0.0.0")),
        port: env::var("PORT").unwrap_or_else(|_| String::from("8080")).parse()?,
        existdb_uri: String::from("http://localhost:8080/exist/rest/dmarc"),
        auth_userheader: String::from("remote-user"),
        dev_authuser: None,
    })
}

File: webapp-go/appconfig.go
package main

import (
  "errors"
  "os"
  "strconv"
)

type appConfig struct {
  loglevel string
  host string
  port uint16
  existdbUri string
  authUserHeader string
  authGroupsHeader string
  devAuthuser *string
}

func (cfg appConfig) init() appConfig {
  cfg.loglevel = "info"
  cfg.host = "localhost"
  cfg.port = 8080
  cfg.existdbUri = "http://localhost:8081/exist/rest/dmarc"
  cfg.authUserHeader = "remote-user"

  return cfg
}

func ReadAppConfig() (appConfig, error) {
  cfg := appConfig{}.init()

  if val, ok := os.LookupEnv("LOGLEVEL"); ok {
    cfg.loglevel = val
  }
  if val, ok := os.LookupEnv("HOST"); ok {
    cfg.host = val
  }
  if val, ok := os.LookupEnv("PORT"); ok {
    if ival, err := strconv.Atoi(val); err == nil {
      cfg.port = uint16(ival)
    } else {
      return appConfig{}, errors.New("could not parse value of env var PORT")
    }
  }
  if val, ok := os.LookupEnv("EXISTDB_URI"); ok {
    cfg.existdbUri = val
  }
  if val, ok := os.LookupEnv("AUTHUSER_HEADER"); ok {
    cfg.authUserHeader = val
  }
  if val, ok := os.LookupEnv("AUTHGROUPS_HEADER"); ok {
    cfg.authGroupsHeader = val
  }
  if val, ok := os.LookupEnv("DEV_AUTHUSER"); ok {
    cfg.devAuthuser = &val
  }

  return cfg, nil
}

